### ISSUE TITLE
Suggestion Excel Quit Event

### ISSUE THREAD

**Issue by caioproiete** [sandboxorg/exceldnaissuestest#32]
More of a suggestion rather than a direct issue. 

I know that a few people have mentioned that it is hard to know exactly when Excel is quitting permanently (e.g. the AutoClose issue). The current work around is to have a COM addin that signals this OnDestroy().

How about using the SynchronizationWindow class that is already in there to raise an event in the add-in when the window receives a WM_DESTROY message?

Something along the lines of:
public const int WM_DESTROY = 0x0002; // Sent before child windows are destroyed
public const int WM_NCDESTROY = 0x0082; // Sent after child windows have been destroyed      

public event EventHandler Quit;

protected override void WndProc(ref System.Windows.Forms.Message m)
{
if (m.Msg == WM_DESTROY)
    Quit(this, EventArgs.Empty);

base.WndProc(ref m);

}

Just a thought... :)

#### This work item was migrated from CodePlex
CodePlex work item ID: '8894'
Vote count: '1'

**Comment by caioproiete** [comment]
[govert@9/5/2013]
I've added a more detailed example of the approach of using a COM Add-In to detect the shutdown here: See the discussion here: https://exceldna.codeplex.com/wikipage?title=AutoClose%20and%20Detecting%20Excel%20Shutdown

One concern with your suggestion based on the windows message, is that we might not be able to interact with the Excel object model in that context, and further code that depends on the windows message processing might also be affected, like showing a model dialog box.

**Comment by caioproiete** [comment]
[sverrirs@9/5/2013]
Hi Govert,

Thanks for the response. 

I attempted to use the COM approach recently with Excel 2010 but had to abandon it as: 

1. The OnBeginShutdown was never called for my COM addin2. The OnDisconnection was called when the last open workbook in Excel was closed (but Excel was still running)

Strange behavior and I couldn't for the life of me get around it so I had to adopt the WM_DESTROY method above. 

I can agree that this approach might not be generic enough to be safe for everyone to use and therefore not a good candidate for the library.

Keep up the good work,Sverrir S.
